src/core/layertree/qgslayertreemodel.cpp

```diff
@@ -307,7 +307,7 @@ QVariant QgsLayerTreeModel::data( const QModelIndex &index, int role ) const
     if ( QgsLayerTree::isLayer( node ) )
     {
       const QgsMapLayer* layer = QgsLayerTree::toLayer( node )->layer();
-      if ( layer && !layer->isInScaleRange( mLegendMapViewScale ) )
+      if ( !node->isVisible() || ( layer && !layer->isInScaleRange( mLegendMapViewScale ) ) )
```

Comment:
I suspect that when a child node changes color due to a parent update, it is not correctly redrawn. The redrawing only happens when passing the curser. See more visual description here: https://www.youtube.com/watch?v=j4eyQicYAOc